server/src/main/java/org/apache/druid/curator/CuratorModule.java

```diff
@@ -127,6 +154,29 @@ public EnsembleProvider makeEnsembleProvider(CuratorConfig config, ExhibitorConf
       return new FixedEnsembleProvider(config.getZkHosts());
     }
 
+    RetryPolicy retryPolicy;
+    if (config.getTerminateDruidProcess()) {
+      final Function<Void, Void> exitFunction = new Function<Void, Void>()
+      {
+        @Override
+        public Void apply(Void aVoid)
+        {
+          log.error("Zookeeper can't be reached, forcefully stopping virtual machine...");
+          System.exit(1);
```

Comment:
```java
try {
  log.error("Zookeeper can't be reached, forcefully stopping virtual machine...");
} finally {
  System.exit(1);
}
```

Reply:
Switched to this pattern.  I didn't think it was necessary since I've never seen a logger kill a JVM.

Reply:
Should this stop the lifecycle too? If this is getting hit it means that the curator killer is _not_ getting hit and so I think this exit would be unclean.

Also, did you figure out why this needs to kill the process too? I see the discussion here, https://github.com/apache/incubator-druid/pull/6740#discussion_r243368737 but it doesn't really come to a conclusion about whether it's needed. 

I'm not sure it is needed. Presumably if exhibitor is down the zk processes managed by it are dead too, and I would expect the other killer will get hit? Or is this for a startup case where the other killer somehow doesn't get triggered? Regardless, could you also add a comment with the explanation?

Reply:
@clintropolis I think the problem is that nobody knows answers to your questions and nobody has the determination to invest the effort to research this question (yet). I don't think we should demand this from @michael-trelinski. I think the PR in its current form is a move in the right direction anyway.

However, @michael-trelinski could you please add a comment in `makeEnsembleProvider()` along the lines with ```// It's unknown whether this precaution is needed, see https://github.com/apache/incubator-druid/pull/6740#discussion_r243368737 for details```?

Reply:
Sure, I wasn't expecting a deep investigation, I was suggesting either adding a `lifecycle.stop()` to make sure it stops cleanly and adding a comment to explain the we aren't sure if it is needed, OR just removing it until a scenario where an Exhibitor run Zookeeper is down but Druid doesn't kill itself and adding this block later when we find out it is needed.

Reply:
Ok, have looked at this a bit more, and I think it might not matter if `lifecycle.stop` is called here (or in the other killer) because _I think_ it will probably get picked up and called by the shutdown hook if `System.exit(1)` is called directly here. 

I think it would be nice if these retry killer logic blocks were self consistent at least, either both calling `lifecycle.stop` explicitly to be sure it gets called, or neither calling it and counting on the shutdown hook, but if @leventov is ok with this as is I don't feel strongly enough to block merging and my review can be counted as a +1

Reply:
`lifecycle` isn't available in the scope of `makeEnsembleProvider`.

Reply:
Arguments to those provider methods are injected, so it could be available if you added it as an argument, e.g. 

```
makeEnsembleProvider(CuratorConfig config, ExhibitorConfig exConfig, Lifecycle lifecycle)
{
...
```

Reply:
In `master` and the branch/commit I'm using, `makeEnsembleProvider` is declared thusly:

`@Provides
@LazySingleton
public EnsembleProvider makeEnsembleProvider(CuratorConfig config, ExhibitorConfig exConfig)`

Reply:
I see what you're getting at.  Apologies.

However, if it wasn't there to begin with and we're all not sure why makeEnsembleProvider is even there, should I be mucking around with it by adding a Lifecycle?

Reply:
I don't think there is question of why `makeEnsembleProvider` exists, it provides the `EnsembleProvider` argument to `makeCurator`. 

The only question I had is whether the quit on retry functionality needs to be wired into it's failures as well as curator's failures. I believe there should be no harm in using the injected lifecycle here as far as I can tell.

Reply:
@michael-trelinski could you please link this discussion from the comment? There is no superstition that code should be "self-contained" in Druid, there are many links from the code to Github. It would be easier for readers to navigate. Currently, readers would be forced to dig into Github history to follow this path.